### DISCUSSION TITLE
env_file not working (?) in DockerOperator, is this a bug?

### DISCUSSION THREAD

**Discussion by MattiaGallegati** [apache/airflow#29990] [Q&A] [answered]
Hello everyone,
I've recently experienced troubles while using a DockerOperator where I use the env_file parameter to pass an env file from the host machine (in my case the host machine is the machine of a CeleryWorker).
If I print the environment variables using "printenv" inside the DockerOperator command it seems that the only variables that are loaded inside the container are the one coming from my base image/placed by the Dockerfile.
I use a simple absolute path to gather the env file (e.g. /home/xyz/file.env), or a Jinja templated variables value access.

My architecture is Airflow 2.5.0, with a Master node and a worker node, with the latest docker-provider. 

Is there something I'm missing or is this a bug?
Thank you in advace.

**Comment by potiuk** [comment]
You might also be doing something wrong. Hard to say as you have not provided detailed example of what you do and how you do it.

**Comment by Taragolis** [comment, marked as answer]
I guess all problem refers to "interesting" implementation of `env_file` in DockerOperator.

Last time when I have a look on this operator the idea of usage was use Airflow `rendered_fields` behaviour.

If you provide path:
   - By default path to files should relative to DAG directory, and could extended by `template_undefined` argument in [DAG](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/models/dag/index.html#airflow.models.dag.DAG). This built in top of [`jinja2.FileSystemLoader`](https://jinja.palletsprojects.com/en/3.0.x/api/#jinja2.FileSystemLoader)
   - Filename should have extension `.env`, e.g. you path should be `foo/bar/awesome.env`
   - No jinja syntax in path, e.g.  you can't use `foo/bar/{{ var.value.some_variable }}.env`

You could also provide.... a values to this argument, but due to implementation it expected string, something like

```python
ENV = """
VAR1=1
VAR2=2
VAR3={{ ds }}
"""

...

DockerOperator(
    task_id="foo-bar",
    ...
    env_file=ENV
)
```

I hope this information help you.

**Reply by MattiaGallegati** [reply]
Thanks to your help I ""figured out"" the implementation of this parameter.
I saw this thing working as expected when providing:
- Top level variable in the form of string as you indicated.
- File .env when it's placed in the DAG folder e.g. by passing _file.env_ .

Btw, it can't figure out how to make this works in the following cases:
- .env placed at an higher level (e.g in a folder backward from the DAG folder). I've tryed absolute path or "relative dotted" path _/home/user/file.env_ or _../file.env_
- .env file placed in a subfolder of the DAG folder e.g. _my_folder/file.env_
- Jinja templated paths. If I create a variable containing the value file.env and call it in the form of _{{ var.value.my_env_path }}_ it will not work. Strange as this is a templated field.

Do you have any idea about that?
Btw, is this actually the expected behaviours?

**Reply by Taragolis** [reply]
> .env placed at an higher level (e.g in a folder backward from the DAG folder). I've tryed absolute path or "relative dotted" path /home/user/file.env or ../file.env

This would not work, that is how [jinja2.FileSystemLoader](https://jinja.palletsprojects.com/en/3.0.x/api/#jinja2.FileSystemLoader) work. All absolute paths use as relative to search paths, relative path searches in current working directory (AFAIK it is DAG dir)

You could set search path in DAG level, e.g.:

```python
with DAG(dag_id="awesome_dag", ..., template_searchpath=["/home/user"]):
    DockerOperator(
        task_id="foo-bar",
        ...
        env_file="/file.env"
    )
```

> .env file placed in a subfolder of the DAG folder e.g. my_folder/file.env

This should work out of the box

```python
    DockerOperator(
        task_id="foo-bar",
        ...
        env_file="/my_folder/file.env"
    )
```

> Jinja templated paths. If I create a variable containing the value file.env and call it in the form of {{ var.value.my_env_path }} it will not work. Strange as this is a templated field.

If I do not miss and this feature not implemented than it not possible for templating in Airflow use either filepath or value, but filepath can't contain Jinja templates

**Reply by MattiaGallegati** [reply]
You are right.
The env_file parameter is "space-sensitive" so blank spaces can create problem in the path parsing.

This parameter definetly is not as easy to use and is not well documented. It's almost impossible to figure out all this "tricks" to make this works without deep diving in the code.
Hope that someone extends the documentation, or maybe I will do it in the future.

Thank you so much for you precious help.
